core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ReplicationStrategy.java

```diff
@@ -17,11 +17,10 @@
 
 import com.datastax.oss.driver.api.core.metadata.Node;
 import com.datastax.oss.driver.api.core.metadata.token.Token;
-import com.datastax.oss.driver.shaded.guava.common.collect.SetMultimap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 public interface ReplicationStrategy {
-  SetMultimap<Token, Node> computeReplicasByToken(
-      Map<Token, Node> tokenToPrimary, List<Token> ring);
+  Map<Token, Set<Node>> computeReplicasByToken(Map<Token, Node> tokenToPrimary, List<Token> ring);
```

Comment:
Incidentally, I find the new return type more expressive than the old one.

Reply:
By the way, in `KeyspaceTokenMap`: shouldn't we also migrate `tokenRangesByNode` to `Map<Node, Set<TokenRange>>`? I'm not a big fan of multimaps.

Reply:
A multimap is more convenient in `KeyspaceTokenMap.buildTokenRangesByNode`: unlike in replication strategies, we don't build the sets one at a time. It's essentially an inversion of `replicasByToken`, so as we iterate each token we dispatch it to the "in progress" set of each replica. We would have to use a temporary map of set builders, and then build all the builders in a second pass at the end. Then we could try to also optimize duplicates, although I'm not sure it's common (or even possible?) that two nodes replicate the exact same set of tokens.

But another question is whether we should precompute that information at all. `replicasByToken` is where we really want speed, because it's what the LBP calls for token-aware routing. But `tokenRangesByNode` is only used when the client application calls `TokenMap.getTokenRanges(CqlIdentifier, Node)`. We could trade space for speed and recompute the result on every call, it "only" costs a full traversal of `replicasByToken`, which is O(ring_size * RF). Or maybe initialize `tokenRangesByNode` lazily...

Maybe we can leave that as a follow-up ticket because I need to refresh my memory on why we added this method and how it's typically used in the field. There's some background on [JAVA-312](https://datastax-oss.atlassian.net/browse/JAVA-312) on that.

Reply:
I created [JAVA-2874](https://datastax-oss.atlassian.net/browse/JAVA-2874).